.travis.yml

```diff
@@ -95,7 +95,7 @@ install:
     - pip install -U pip
     - if [[ "${MAIN_CMD}" == pycodestyle ]]; then pip install pycodestyle; fi
     - if [[ "${MAIN_CMD}" == sphinx-build* ]]; then pip install Sphinx; fi
-    - if [[ "${MAIN_CMD}" == pytest ]]; then pip install -r requirements.txt; fi
+    - if [[ "${MAIN_CMD}" == pytest ]]; then pip install -r requirements.txt; pip install . ; fi
```

Comment:
This seems like a very sneaky change.  Could you please explain in detail why this is necessary?  Test were working perfectly fine without this prior to adding just one module.

Reply:
Previously, the unit tests were running "in place" from setup.py.  This change installs the package first before running the tests.  Many packages do this in order to test that the installation process produces a working, installed package.  Example:

https://github.com/desihub/fiberassign/blob/master/.travis.yml#L135

In this particular instance, I am generating and running external python scripts to test the redirection, because this redirection does not work inside the python unittest framework (the unittest runner has a handle to stdout during the tests, and redirecting it invalidates that handle, raising an I/O exception).

In order to have an external script run the redirection, desiutil must be installed and importable.  That is the reason behind this change.

Reply:
The problem with that in the future we will not be using `setup.py` for testing anymore.  In fact, desiutil is already using `pytest` instead, and we will probably eventually move to `tox`.  So I don't think this change is future-proof.

Reply:
If we were using a tool like `tox`, then the package would be automatically installed before running the tests (see https://tox.readthedocs.io/en/latest/#current-features), and so this workaround would not be needed.  It is only needed because we are running the tests using `setup.py`.

Let me know if there are any other changes needed.  I think I have resolved all your previous review points.  The fiberassign commissioning script is ready to use this function for its logging once it is merged.

Reply:
OK, but again, we are not using setup.py for tests now. pytest is used instead. I am pretty sure that the travis tests already reflect this. Where are the remaining tests being run that still use python setup.py test?

Reply:
Sure.  pytest best practices (https://docs.pytest.org/en/latest/goodpractices.html) recommend installing the package in editable mode first.  In our case we are running tests once in a CI instance, so installing in normal mode is fine.  I'm confused what changes you are looking for here?  

From my perspective, testing against the installed package has the benefit of testing the install process as well.  And as previously noted, for this type of redirection there is no other way to test it.

Reply:
Maybe we're talking past each other here, and should move this to an actual audio conversation.  But let me try one more time to make myself clear.

* Travis tests were already working *with* `pytest` and *without* this change.
* You have not shown me any cases where `python setup.py test` is still being used.
* Therefore, I do not understand why this change is necessary.  Best practice or no best practice.
* This PR is supposed to be about this redirect module, not changing the nature of the Travis tests.

Reply:
1. `pytest` is currently being used *without* installing the package first.  The earlier statement about `setup.py` was an error.  The point is that the package is not being installed first before running the tests.

2.  Quoting my earlier statement in this same thread:  _"I am generating and running external python scripts to test the redirection, because this redirection does not work inside the python unittest framework ...  In order to have an external script run the redirection, desiutil must be installed and importable. That is the reason behind this change."_

3.  I am only changing the travis file to install the package first, as required in order to test the redirection function.

4.  My statements / links about "best practices" were simply to counter your seeming resistance to installing the package prior to testing.  I was just trying to communicate that such a step is a perfectly reasonable thing to do and is already done automatically with some test runners like `tox`.

Reply:
> desiutil must be installed and importable

Then why not set PYTHONPATH?